File: src/python/challenges/problems/find_minimum_in_rotated_sorted_array_ii_test.py
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii

from typing import List
from core.problem_base import *

class FindMinimumInRotatedSortedArrayIi(ProblemBase):
    def Solution(self, nums: List[int]) -> int:
        left = 0
        right = len(nums) - 1
        
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            elif nums[left] == nums[mid] and nums[mid] == nums[right]:
                right -= 1
            else:
                right = mid

        return nums[left]


if __name__ == '__main__':
    TestGen(FindMinimumInRotatedSortedArrayIi) \
        .Add(lambda tc: tc.Param([10,1,10,10,10]).Result(1)) \
        .Add(lambda tc: tc.Param([3,3,3,5,1,3,3]).Result(1)) \
        .Add(lambda tc: tc.Param([3,3,1,3]).Result(1)) \
        .Add(lambda tc: tc.Param([1,3,3]).Result(1)) \
        .Add(lambda tc: tc.Param([1,3,5]).Result(1)) \
        .Add(lambda tc: tc.Param([2,2,2,0,1]).Result(0)) \
        .Add(lambda tc: tc.Param([1,1,1,1,1,1,1,1,1,1,1,1,1]).Result(1)) \
        .Run()

File: src/go/challenges/problems/find_minimum_in_rotated_sorted_array_ii_test.go
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type FindMinimumInRotatedSortedArrayIi struct{}

func TestFindMinimumInRotatedSortedArrayIi(t *testing.T) {
	gen := core.TestSuite[FindMinimumInRotatedSortedArrayIi]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{10, 1, 10, 10, 10}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{3, 3, 3, 5, 1, 3, 3}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{3, 3, 1, 3}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 3, 3}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 3, 5}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 1, 1, 1, 1}).Result(1)
	}).Run(t)
}

func (FindMinimumInRotatedSortedArrayIi) Solution(nums []int) int {
	left := 0
	right := len(nums) - 1

	for left < right {
		mid := (left + right) / 2
		if nums[mid] > nums[right] {
			left = mid + 1
		} else if nums[left] == nums[mid] && nums[mid] == nums[right] {
			right--
		} else {
			right = mid
		}
	}

	return nums[left]
}
